ktor-client/ktor-client-features/ktor-client-auth/common/src/io/ktor/client/features/auth/Auth.kt

```diff
@@ -41,11 +41,24 @@ public class Auth(
                 if (origin.request.attributes.contains(circuitBreaker)) return@intercept origin
 
                 var call = origin
-                val candidateProviders = HashSet(feature.providers).apply { removeAll(feature.alwaysSend) }
+                val candidateProviders = HashSet(feature.providers).apply {
+                    removeAll(feature.alwaysSend - feature.providers.find { it is BearerAuthProvider })
+                }
+
                 while (call.response.status == HttpStatusCode.Unauthorized) {
-                    val headerValue = call.response.headers[HttpHeaders.WWWAuthenticate] ?: return@intercept call
+                    val headerValue = call.response.headers[HttpHeaders.WWWAuthenticate]
+                    if (headerValue.isNullOrEmpty()) {
+                        return@intercept call
+                    }
+
                     val authHeader = parseAuthorizationHeader(headerValue) ?: return@intercept call
                     val provider = candidateProviders.find { it.isApplicable(authHeader) } ?: return@intercept call
+                    when(provider) {
+                        is BearerAuthProvider -> {
+                            provider.refreshToken(call) ?: return@intercept call
```

Comment:
Should we block other requests while token refresh is in progress?

Reply:
Fixed